model/encoder.py

```diff
@@ -0,0 +1,242 @@
+import torch
+import torch.nn as nn
+from attention import MultiHeadAxialAttention
+
+
+class ConvLayer(nn.Module):
+    """
+    This class implements the convolution layer
+    of the encoder block.
+
+    Link to RNAinformer conv layer if you want to see what
+    they did:
+    https://anonymous.4open.science/r/RNA-design-7204/RNAinformer/modules/feed_forward.py
+    """
+
+    def __init__(self, embed_size: int, kernel_size: int = 3,
+                 out_channels: int = 256, padding="same"):
+        """
+        Arguments:
+            - embed_size: embedding size of the model,
+                          represents the input channel number
+            - kernel_size: kernel size of the convolution
+            - out_channels: number of out channels of the convolution
+            - padding: type of padding, preferably always same
+        """
+        super(ConvLayer, self).__init__()
+
+        """
+        RNAInformer uses GroupNorm, will have some questions
+        """
+        self.norm1 = nn.GroupNorm(1, embed_size)
+        self.conv1 = nn.Conv2d(in_channels=embed_size,
+                               out_channels=out_channels,
+                               kernel_size=kernel_size, padding=padding)
+        # SiLU is used in RNAInformer, so why not try?
+        self.act1 = nn.SiLU()
+
+        # they also didnt use second normalisation, so i didnt too
+        self.conv2 = nn.Conv2d(in_channels=out_channels,
+                               out_channels=embed_size,
+                               kernel_size=kernel_size, padding=padding)
+        """
+        RNAInformer doesnt use activation function after second layer,
+        I didnt like it, so I used it
+        """
+        self.act2 = nn.SiLU()
+
+    def forward(self, x):
+        """
+        I added residual connection to mimic what we did in last semester,
+        they also do residuals
+        """
+        """
+        For convolution to work, need to reshape data from
+        (bs, N, N, embedd) to
+        (bs, embedd, N, N), also, needed to use 2d conv,
+        where embedding dim is actually channels??,
+        whats the deal with that, is there some
+        paper or article which explains this in more
+        detail, or is this one of it works, we dont know why things
+        in ML
+        """
+        _x = x
+        x = x.permute(0, 3, 1, 2)
+        # does normalisation go before or after i remeber residual connections?
+        x = self.norm1(x)
```

Comment:
Some people use it before, some after. Doubt it will affect results too much. Leave it like it is for now, we can easily try changing it later.